core/src/test/java/org/apache/iceberg/TestTableMetadata.java

```diff
@@ -1687,6 +1687,44 @@ public void testV3TimestampNanoTypeSupport() {
         3);
   }
 
+  @Test
+  public void testV3VariantTypeSupport() {
```

Comment:
I know this is copying a lot of tests in this class but we should start future proofing a bit more imho. We also have tests around this sort of thing in TestSchema.java. I think it is probably ok to just keep all of our schema validation tests there, but it wouldn't hurt to have some redundancy here as well.

Refactoring the whole suite can come in another pr but I think we should build a templated test that's something like


```java
@ParameterizedTest
@ValueSource(types = {Types.TimetstampNanos, Types.Variant, ....})
testTypeSupport(Type type) {
  Schema schemaWithType =  new Schema(
          Types.NestedField.required(1, "id", Types.LongType.get()),
          Types.NestedField.optional(2, type.name, type),
          Types.NestedField.optional(3, "arr", Types.ListType.ofRequired(4, type)),
          Types.NestedField.required(5, "struct", 
            Types.StructType.of(
                  Types.NestedField.optional(6, "inner_" + type.name, type),
                  Types.NestedField.required(7, "data", Types.StringType.get()))),
          Types.NestedField.optional(8, "struct_arr",
              Types.StructType.of(
                  Types.NestedField.optional(9, "ts", type))));
     
    //Psuedo code here 
    from 0 -> MIN_FORMAT_VERSION
       fail to make metadata
       
    from MIN_FORMAT_VERSION -> SUPPORTED_TABLE_VERSION
       succeed
}

The most important part about this is that we wouldn't have to continually update tests every time a new valid metadata version is added. It also would be much easier to test type compatibility. (I'm thinking that Geo is going to need the exact same thing soon)

Reply:
Not this is part of my goal to remove all tests that have V3 or V2 in their title.

Reply:
Also todo add Variant to Schema.MIN_FORMAT_VERSION

Reply:
Thanks @RussellSpitzer  for already creating a [PR](https://github.com/apache/iceberg/pull/11444) to address that. I will include that when it's merged.

Reply:
+1 for moving this to `TestSchema`. The test here should really just be validating that schemas are validated against the table version, but the actual validation tests should be in `TestSchema`.

Reply:
Lets just drop this test then (or add a single variant field into the above test) and just change it to be called testV3SchemaValidation or something

Reply:
I removed the actual validation for both variant and timestamp_ns since it's already covered in TestSchema. I added version check/validation coverage for TableMetadata.newTableMetadata().